server/service/queries.go

```diff
@@ -287,10 +292,23 @@ func (svc *Service) ModifyQuery(ctx context.Context, id uint, p fleet.QueryPaylo
 	}
 	if p.Logging != nil {
 		query.Logging = *p.Logging
+		if *p.Logging != fleet.LoggingSnapshot {
+			if err := svc.ds.DeleteAllResultsForQuery(ctx, id); err != nil {
+				return nil, err
+			}
+		}
 	}
 	if p.ObserverCanRun != nil {
 		query.ObserverCanRun = *p.ObserverCanRun
 	}
+	if p.DiscardData != nil {
+		query.DiscardData = *p.DiscardData
+		if *p.DiscardData == true {
+			if err := svc.ds.DeleteAllResultsForQuery(ctx, id); err != nil {
```

Comment:
As is, it seems it could execute `svc.ds.DeleteAllResultsForQuery` multiple times (e.g. if changing both the SQL and the logging to differential)

Let's define a `var deleteResults bool` at the beginning and call `svc.ds.DeleteAllResultsForQuery` once before (or after?) `svc.ds.SaveQuery`.